test/v2/client/test_client.go

```diff
@@ -513,72 +542,147 @@ func (c *TestClient) DispersePayload(ctx context.Context, payloadBytes []byte) (
-		}
+	start := time.Now()
 
-		payloadBytesFromRelay := payload.Serialize()
+	if timeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, timeout)
+		defer cancel()
+	}
 
-		if !bytes.Equal(payloadBytesFromRelay, expectedPayload) {
-			return fmt.Errorf("payloads do not match")
-		}
+	blobBytesFromRelay, err := c.relayClient.GetBlob(ctx, relayKey, key)
+	if err != nil {
+		return fmt.Errorf("failed to read blob from relay: %w", err)
+	}
+
+	c.metrics.reportRelayReadTime(time.Since(start), relayKey)
+
+	blob, err := coretypes.DeserializeBlob(blobBytesFromRelay, blobLengthSymbols)
+	if err != nil {
+		return fmt.Errorf("failed to deserialize blob: %w", err)
+	}
+
+	payload, err := blob.ToPayload(c.payloadClientConfig.PayloadPolynomialForm)
+	if err != nil {
+		return fmt.Errorf("failed to decode blob: %w", err)
+	}
+
+	payloadBytesFromRelay := payload.Serialize()
+
+	if !bytes.Equal(payloadBytesFromRelay, expectedPayload) {
+		return fmt.Errorf("payloads do not match")
 	}
 
 	return nil
 }
 
+// GetCurrentBlockNumber retrieves the current block number from the indexed chain state.
+func (c *TestClient) GetCurrentBlockNumber(ctx context.Context) (uint64, error) {
+	currentBlockNumber, err := c.indexedChainState.GetCurrentBlockNumber(ctx)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get current block number: %w", err)
+	}
+	return uint64(currentBlockNumber), nil
+}
+
 // ReadBlobFromValidators reads a blob from the validators and compares it to the given payload.
+//
+// The timeout provided is a timeout for each read from a quorum, not all reads as a whole.
 func (c *TestClient) ReadBlobFromValidators(
 	ctx context.Context,
 	blobKey corev2.BlobKey,
 	blobVersion corev2.BlobVersion,
 	blobCommitments encoding.BlobCommitments,
 	quorums []core.QuorumID,
-	expectedPayloadBytes []byte) error {
+	expectedPayloadBytes []byte,
+	timeout time.Duration,
+	validateAndDecode bool) error {
 
 	currentBlockNumber, err := c.indexedChainState.GetCurrentBlockNumber(ctx)
 	if err != nil {
 		return fmt.Errorf("failed to get current block number: %w", err)
 	}
 
 	for _, quorumID := range quorums {
-		c.logger.Debugf("Reading blob from validators for quorum %d", quorumID)
+		err = c.ReadBlobFromValidatorsInQuorum(
+			ctx,
+			blobKey,
+			blobVersion,
+			blobCommitments,
+			quorumID,
+			uint64(currentBlockNumber),
+			expectedPayloadBytes,
+			timeout,
+			validateAndDecode)
+		if err != nil {
+			return fmt.Errorf("failed to read blob from validators in quorum %d: %w", quorumID, err)
+		}
+	}
 
+	return nil
+}
+
+// ReadBlobFromValidatorsInQuorum reads a blob from the validators in a specific quorum and compares it to
+// the given payload.
+func (c *TestClient) ReadBlobFromValidatorsInQuorum(
+	ctx context.Context,
+	blobKey corev2.BlobKey,
+	blobVersion corev2.BlobVersion,
+	blobCommitments encoding.BlobCommitments,
+	quorumID core.QuorumID,
+	currentBlockNumber uint64,
```

Comment:
this should be `RBN`, not `currentBlockNumber`

Reply:
fixed